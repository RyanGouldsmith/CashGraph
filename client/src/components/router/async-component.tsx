import React from 'react';

interface DynamicComponentInterface {
  [key: string]: React.ComponentClass<{}, {}> | React.SFC<{}>;
}

interface ComponentState {
  Component: null | React.ComponentClass<{}, {}> | React.SFC<{}>;
}

export const AsyncComponent = (
  loader: () => Promise<DynamicComponentInterface>,
  componentName: string,
) =>
  class AsyncComp extends React.Component<{}, ComponentState> {
    constuctor() {
      this.state = {
        Component: null,
      };
    }

    public async componentDidMount() {
      const ComponentItem = await loader();
      this.setState({ Component: ComponentItem[componentName] });
    }

    public render() {
      const Component = this.state && this.state.Component;
      return Component ? <Component {...this.props} /> : null;
    }
  };
