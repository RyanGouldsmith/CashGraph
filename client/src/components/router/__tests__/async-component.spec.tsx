import 'babel-polyfill';
import React from 'react';
import { cleanup, render, wait } from 'react-testing-library';

import { AsyncComponent } from '../async-component';

afterEach(cleanup);

test('Should render a component based on a promise', async () => {
  const loader = () =>
    Promise.resolve({
      'a-component': () => <div>A Test</div>,
    });
  const AysncComp = AsyncComponent(loader, 'a-component');
  const renderedAsyncComp = render(<AysncComp />);

  await wait(() => {
    expect(renderedAsyncComp).toMatchSnapshot();
  });
});

test('Should return null when the component name does not match the loader', async () => {
  const loader = () =>
    Promise.resolve({
      'A-Component': () => <div>A Test</div>,
    });
  const AysncComp = AsyncComponent(loader, 'NotHere');
  const renderedAsyncComp = render(<AysncComp />);

  await wait(() => {
    expect(renderedAsyncComp).toMatchSnapshot();
  });
});
