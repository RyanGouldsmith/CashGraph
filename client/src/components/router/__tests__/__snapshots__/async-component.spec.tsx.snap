// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render a component based on a promise 1`] = `
Object {
  "baseElement": <body>
    <div>
      <div>
        A Test
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      A Test
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Should return null when the component name does not match the loader 1`] = `
Object {
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
